### PR TITLE
Platform assignments feedback

### PR DESCRIPTION
**Pull request by barbeau**
Changes from the current google/transit PR:
* Name change of `platform_id` -> `assigned_stop_id` - The assigned stop may not be a platform within a station. Other potential naming options include `stop_id`, `new_stop_id`, `close_stop_id`, `adjacent_stop_id`, `neighboring_stop_id`, `?`.
* Add more subjective language to define when `assigned_stop_id` can be used (versus requiring changes to be only within a station)
* Require that `StopTimeUpdate.stop_sequence` be used as the selector and that `StopTimeUpdate.stop_id` be empty
* Add that you can assign a stop without providing real-time predictions by using `StopTimeUpdate.schedule_relationship = NO_DATA`
* Don't block use of `StopTimeUpdate.stop_id` being populated with other values in case others want to have unofficial GTFS-realtime feeds that changes `stop_id` to represent a different stop (e.g., Bliksem Labs B.V.)

### CODE REVIEW

File: gtfs-realtime/spec/en/reference.md [hunk 1]
```diff
@@ -199,6 +201,18 @@ The relation between this StopTime and the static schedule.
 | **NO_DATA** | No data is given for this stop. It indicates that there is no realtime information available. When set NO_DATA is propagated through subsequent stops so this is the recommended way of specifying from which stop you do not have realtime information. When NO_DATA is set neither arrival nor departure should be supplied. |
 | **UNSCHEDULED** | The vehicle is operating a frequency-based trip (GTFS frequencies.txt with exact_times = 0). This value should not be used for trips that are not defined in GTFS frequencies.txt, or trips in GTFS frequencies.txt with exact_times = 1. Trips containing `stop_time_updates` with `schedule_relationship: UNSCHEDULED` must also set the TripDescriptor `schedule_relationship: UNSCHEDULED` <br>**Caution:** this field is still **experimental**, and subject to change. It may be formally adopted in the future.<br>.
 
+## _message_ StopTimeProperties
+
+Realtime update for certain properties defined within GTFS stop_times.txt.
+
+**Caution:** this message is still **experimental**, and subject to change. It may be formally adopted in the future.<br>.
```
Comment (timMillet) [inline gtfs-realtime/spec/en/reference.md:208]: `.<br>` instead of `.<br>.`

File: gtfs-realtime/spec/en/reference.md [hunk 2]
```diff
@@ -199,6 +201,18 @@ The relation between this StopTime and the static schedule.
...
...elopers.google.com/protocol-buffers/docs/proto#scalar) | Optional | One | Supports real-time stop assignments. Refers to a stop_id defined in the GTFS stops.txt. <br> The new `assigned_stop_id` should not result in a significantly different trip experience for the end user than the stop_id defined in GTFS stop_times.txt. In other words, the end user should not view this new stop_id as an "unusual change" if the new stop was presented within an app without any additional context. For example, this field is intended to be used for platform assignments by using a stop_id that belongs to the same station as the stop originally defined in GTFS stop_times.txt. <br> To assign a stop without providing any real-time arrival or departure predictions, populated this field and set StopTimeUpdate.schedule_relationship = NO_DATA. <br> If this field is populated, StopTimeUpdate `stop_sequence` must be populated and StopTimeUpdate `stop_id` should not be populated. Stop assignments should be reflected in other GTFS-realtime fields as well (e.g., `VehiclePosition.stop_id`). <br>**Caution:** this field is still **experimental**, and subject to change. It may be formally adopted in the future.<br>. |
```
Comment (timMillet) [inline gtfs-realtime/spec/en/reference.md:214]: missing code text formatting for several `stop_id`

File: gtfs-realtime/spec/en/reference.md [hunk 3]
```diff
@@ -180,11 +181,12 @@ The update is linked to a specific stop either through stop_sequence or stop_id,
...
-| **stop_id** | [string](https://developers.google.com/protocol-buffers/docs/proto#scalar) | Conditionally required | One | Must be the same as in stops.txt in the corresponding GTFS feed. Either stop_sequence or stop_id must be provided within a StopTimeUpdate - both fields cannot be empty. |
+| **stop_sequence** | [uint32](https://developers.google.com/protocol-buffers/docs/proto#scalar) | Conditionally required | One | Must be the same as in stop_times.txt in the corresponding GTFS feed.  Either stop_sequence or stop_id must be provided within a StopTimeUpdate - both fields cannot be empty.  stop_sequence is required for trips that visit the same stop_id more than once (e.g., a loop) to disambiguate which stop the prediction is for. If `StopTimeProperties.assigned_stop_id` is populated, then `stop_sequence` must be populated and `stop_id` must *not* be populated. |
```
Comment (timMillet) [inline gtfs-realtime/spec/en/reference.md:184]: maybe removing the part that doesn't concern the field?
"and `stop_id` must *not* be populated"

File: gtfs-realtime/spec/en/reference.md [hunk 4]
```diff
@@ -180,11 +181,12 @@ The update is linked to a specific stop either through stop_sequence or stop_id,
...
+| **stop_id** | [string](https://developers.google.com/protocol-buffers/docs/proto#scalar) | Conditionally required | One | Must be the same as in stops.txt in the corresponding GTFS feed. Either stop_sequence or stop_id must be provided within a StopTimeUpdate - both fields cannot be empty. If `StopTimeProperties.assigned_stop_id` is populated, then `stop_sequence` must be populated and `stop_id` must *not* be populated. |
```
Comment (timMillet) [inline gtfs-realtime/spec/en/reference.md:185]: maybe removing the part that doesn't concern the field?
"`stop_sequence` must be populated"

**Review by timMillet — COMMENTED** [review]
Just a few typos, and two minor questions.

I'm not sure as well of the `assigned_stop_id` field name, but I guess it's the better one from the list, with `new_stop_id`. I'll suggest to use terminology that is already present in the specification, like "update": `stop_id_update` (or, if this naming is reserved for messages, then `updated_stop_id`). Open question to the community?

File: gtfs-realtime/spec/en/reference.md [hunk 5]
```diff
@@ -180,11 +181,12 @@ The update is linked to a specific stop either through stop_sequence or stop_id,
...
 | **departure** | [StopTimeEvent](#message-stoptimeevent) | Conditionally required | One | If schedule_relationship is empty or SCHEDULED, either arrival or departure must be provided within a StopTimeUpdate - both fields cannot be empty. arrival and departure may both be empty when schedule_relationship is SKIPPED.  If schedule_relationship is NO_DATA, arrival and departure must be empty. |
 | **schedule_relationship** | [ScheduleRelationship](#enum-schedulerelationship) | Optional | One | The default relationship is SCHEDULED. |
+| **stop_time_properties** | [StopTimeProperties](#message-stoptimeproperties) | Optional | One | Realtime updates for certain properties defined within GTFS stop_times.txt <br>**Caution:** this field is still **experimental**, and subject to change. It may be formally adopted in the future.<br>. |
```
Comment (gcamp) [inline gtfs-realtime/spec/en/reference.md:189]: Not directly related to this PR but what other fields are we seeing in that object? pickup/dropoff but anything else?

File: gtfs-realtime/proto/gtfs-realtime.proto [hunk 6]
```diff
@@ -242,6 +242,34 @@ message TripUpdate {
...
+    // NOTE: This message is still experimental, and subject to change. It may be formally adopted in the future.
+    message StopTimeProperties {
+      // Supports real-time stop assignments. Refers to a stop_id defined in the GTFS stops.txt.
+      // The new assigned_stop_id should not result in a significantly different trip experience for the end user than
+      // the stop_id defined in GTFS stop_times.txt. In other words, the end user should not view this new stop_id as an
+      // "unusual change" if the new stop was presented within an app without any additional context.
+      // For example, this field is intended to be used for platform assignments by using a stop_id that belongs to the
+      // same station as the stop originally defined in GTFS stop_times.txt.
+      // To assign a stop without providing any real-time arrival or departure predictions, populated this field and set
+      // StopTimeUpdate.schedule_relationship = NO_DATA.
```
Comment (gcamp) [inline gtfs-realtime/proto/gtfs-realtime.proto:255]: 👍

File: gtfs-realtime/spec/en/reference.md (hunk 5 above)
Comment (barbeau) [reply gtfs-realtime/spec/en/reference.md:189]: @gcamp Here are the fields for `stop_time_properties` defined in the [ServiceChanges v3.1 draft spec](http://bit.ly/gtfs-service-changes-v3_1):

~~~
      stop_time_properties {
        assigned_stop_id: "StopId1"
        stop_headsign: "Terminus"
        pickup_type: NO_PICKUP
        drop_off_type: REGULAR_DROP_OFF
        shape_dist_traveled: 514
      }
~~~

See example in the doc [here](https://docs.google.com/document/d/1oPfQ6Xvui0g3xiy1pNiyu5cWFLhvrTVyvYsMVKGyGWM/edit#bookmark=id.46iq4ddo589j).

File: gtfs-realtime/proto/gtfs-realtime.proto [hunk 7]
```diff
@@ -242,6 +242,34 @@ message TripUpdate {
...
+
+      // The extensions namespace allows 3rd-party developers to extend the
+      // GTFS Realtime Specification in order to add and evaluate new features
+      // and modifications to the spec.
+      extensions 1000 to 1999;
+
+      // The following extension IDs are reserved for private use by any organization.
+      extensions 9000 to 9999;
+    }
+    // NOTE: This field is still experimental, and subject to change. It may be formally adopted in the future.
```
Comment (tleboulenge) [inline gtfs-realtime/proto/gtfs-realtime.proto:270]: Add a brief description such as "updated properties for this stop-time"

**Comment by tleboulenge** [comment]
I like "assigned_stop_id", if we can normalize this feature and all language around it to "assignment" (as opposed to a generic "change" - this emphasizes IMO the business-as-usual character of it).

The softened requirement (moved to a comment) makes sense, and gives the explicit example of being in the same station, which we (Google) can then use to announce this is the only case we _currently_ support (but others can ignore).

**Comment by barbeau** [comment]
Closing, the google/transit PR has been updated with all of these changes squashed into a single commit.

**Pull request closed without merging**